### PR TITLE
Convert custom table column configuration from localStorage to user settings

### PR DESCRIPTION
**Pull request by debsmita1**
**Jira Story:**
https://issues.redhat.com/browse/ODC-5110

**Analysis / Root cause**: 
table column configuration stored in local storage should be persisted

**Solution Description**: 
used HOC `withUserSettingsCompatibility` to save configurations in config map

**Screenshots / Gifs for design review**: 
no UI changes

**Unit test coverage report**: 
updated existing test case

**Test setup:**
NA

**Browser conformance**: 

- [x] Chrome
- [x] Firefox
- [ ] Safari
- [ ] Edge

### CODE REVIEW

File: frontend/packages/console-shared/src/hooks/useTableColumns.ts
Comment (debsmita1) [reply frontend/packages/console-shared/src/hooks/useTableColumns.ts]: removed it

File: frontend/packages/console-app/src/components/nodes/NodesPage.tsx
Comment (abhinandan13jan) [inline frontend/packages/console-app/src/components/nodes/NodesPage.tsx:404]: I'm not sure, but do we need a blank `updateTableColumn` or can we return/alter it in the `useTableColumns` definition return statement

File: frontend/packages/console-shared/src/hooks/useTableColumns.ts
Comment (abhinandan13jan) [inline frontend/packages/console-shared/src/hooks/useTableColumns.ts:28]: do we need to return `updateTableColumn` here?

Comment (abhinandan13jan) [inline frontend/packages/console-shared/src/hooks/useTableColumns.ts:28]: nit: loaded is checked twice. Once here and once inside every call in individual components. Can we merge these 2 while we check so that we don't have to hold the loaded value in the component. Then we could also get rid of  statements like
```const [state, , loaded] to const [state];```

**Comment by invincibleJai** [comment]
i tested below scenarios

- namespace table
- project table
- pods table
- node table

Overall it works good , at times i observed flicker when i go to project for example shows shows all columns then update based on what's in Configmap

Nov-30-2020 19-04-57

File: frontend/packages/console-shared/src/hooks/useTableColumns.ts
Comment (christianvogt) [inline frontend/packages/console-shared/src/hooks/useTableColumns.ts:26]: Clients of this hook need to know if the data is still loading:
```suggestion
  const columns = React.useMemo(() => loaded && tableColumns?.[id] || [], [loaded, tableColumns, id]);
  return [columns, updateTableColumn, loaded];
```

File: frontend/packages/console-shared/src/hooks/useTableColumns.ts
Comment (christianvogt) [inline frontend/packages/console-shared/src/hooks/useTableColumns.ts:15]: Can the default value be `undefined` ?

File: frontend/packages/console-shared/src/hooks/useUserSettings.ts
Comment (christianvogt) [inline frontend/packages/console-shared/src/hooks/useUserSettings.ts:101]: The changes in `useUserSettings` will conflict with @sahil143's PR. I suggest you base this PR on his work.

File: frontend/packages/console-shared/src/utils/user-settings.ts
Comment (christianvogt) [inline frontend/packages/console-shared/src/utils/user-settings.ts:65]: This is an async function. You should be throwing errors and returning values instead of returning promises.

File: frontend/public/components/namespace.jsx
Comment (christianvogt) [inline frontend/public/components/namespace.jsx:541]: You need to check if the data has `loaded` otherwise this will flash the default columns first.
Same comment wherever this hook is used.

File: frontend/packages/console-shared/src/hooks/useUserSettings.ts
Comment (debsmita1) [reply frontend/packages/console-shared/src/hooks/useUserSettings.ts:101]: these changes have been copied from Sahil's https://github.com/openshift/console/pull/7298 PR, which he has closed yesterday. I have updated my PR description that it depends on https://github.com/openshift/console/pull/7291.

File: frontend/packages/console-shared/src/utils/user-settings.ts
Comment (debsmita1) [reply frontend/packages/console-shared/src/utils/user-settings.ts:65]: same as above

Comment (sahil143) [reply frontend/packages/console-shared/src/utils/user-settings.ts:65]: I will fix this in my PR. 😄

**Review by sahil143 — CHANGES_REQUESTED** [review]

File: frontend/packages/console-shared/src/hooks/useTableColumns.ts
Comment (christianvogt) [reply frontend/packages/console-shared/src/hooks/useTableColumns.ts]: You want to always maintain stable identity checks on data structures. If you don't you risk triggering dep changes of other hooks.
Of course this is short lived while `!loaded`.
I guess an alternative is to make `[]` a constant instead of using memo.

File: frontend/public/components/namespace.jsx
Comment (christianvogt) [inline frontend/public/components/namespace.jsx:280]: You cannot render the table until the settings have loaded otherwise there's potential for the table to flash the default columns.
Same comment for all usages in this PR.

File: frontend/packages/console-shared/src/hooks/useTableColumns.ts
Comment (christianvogt) [reply frontend/packages/console-shared/src/hooks/useTableColumns.ts:24]: Technically this is correct to have `setTableColumns` as a dependency in case it happens to change.

File: frontend/public/components/namespace.jsx
Comment (debsmita1) [reply frontend/public/components/namespace.jsx]: Done

**Comment by invincibleJai** [comment]
/lgtm

This PR Depends on https://github.com/openshift/console/pull/7291 

Have verified the changes with pod , namespace, node table. Looks good !!

**Comment by christianvogt** [comment]
/hold

waiting on https://github.com/openshift/console/pull/7291

**Comment by christianvogt** [comment]
/approve

will need to rebase after dependent pr merges

**Comment by karthikjeeyar** [comment]
Verified the PR, works fine
/lgtm

**Pull request merged**
